### PR TITLE
[BuildCheck] Add OM and infra for tracking task invocations

### PR DESCRIPTION
**Pull request by ladipro**
Contributes to #9881

### Context

Some build checks will want to analyze tasks executed during build. This PR is adding support for these by introducing several new types and implementing processing of task-related logger events in `BuildEventsProcessor`.

### Changes Made

- Added `RegisterTaskInvocationAction` for analyzers to call to subscribe to task events.
- Added `TaskInvocationAnalysisData` as a unit of reporting task events to analyzers.
- Implemented a transform from `TaskStartedEventArgs`, `TaskFinishedEventArgs`, and `TaskParameterEventArgs` to `TaskInvocationAnalysisData`.

### Testing

New unit tests.

### CODE REVIEW

File: src/Build/BuildCheck/API/IBuildCheckRegistrationContext.cs
Comment (ladipro) [inline src/Build/BuildCheck/API/IBuildCheckRegistrationContext.cs:12]: I am looking for feedback on the name. This could for example be "execution" instead of "invocation": `RegisterTaskExecutionAction`.

File: src/BuildCheck.UnitTests/TaskInvocationAnalysisDataTests.cs
Comment (ladipro) [inline src/BuildCheck.UnitTests/TaskInvocationAnalysisDataTests.cs:141]: This is unfortunate and will very likely require a separate PR to extend `TaskParameterEventArgs` so it carries the name of the parameter as well. Currently it has only the name of the item we're assigning the parameter to.

File: src/MSBuild/XMake.cs
Comment (JanKrivanek) [inline src/MSBuild/XMake.cs:1339]: Note for self: this should ideally be decided on the node after 1st pass ov evaluation and reading of editorconfig. Though it'd as well require the custom analyzers to someho indicate during their registration via the property function what data sources they'll need (to avoid a need to load and register the analyzer in the worker node just for finding it's registrations. And registration in the main node is too late as it happens already after the ProjectEvaluationStartedBuildEventArgs are received - at which point the node can already be running the execution phase)

**Review by JanKrivanek — APPROVED** [review]
This looks good - thank you!!

The only objection I have is about the expressing of the location info.

Btw. - don't you want to extend the `TaskParameterEventArgs` as part of this PR? If not - would be nice to have addition PR for that - possibly merged prior this one.

File: src/Build/BuildCheck/Infrastructure/BuildCheckManagerProvider.cs
Comment (ladipro) [reply src/Build/BuildCheck/Infrastructure/BuildCheckManagerProvider.cs:143]: I don't think I have an instance to call it on. Oh, but acquisition is now in a state that I should be able to register test analyzers directly from project files, is that correct?

File: src/Build/BuildCheck/OM/ParsedItemsAnalysisData.cs
Comment (ladipro) [reply src/Build/BuildCheck/OM/ParsedItemsAnalysisData.cs:61]: > This feels a bit confusing - the location (line, column) is on the base together with the ProjectFile - which seems to guide towards the position being applied to the project file.

It is applied to the project file. It's the line/column where the task is called. Having everything encapsulated in one type would be nicer, I agree, but don't we still have callbacks that are inherently without line/column info like the existing `EvaluatedPropertiesAnalysisData` but are still associated with a file?

Comment (ladipro) [reply src/Build/BuildCheck/OM/ParsedItemsAnalysisData.cs:61]: Aha, sorry, you're pointing out the difference between project file where the particular thing is located vs. project file being built. This makes sense!

File: src/Build/BuildCheck/OM/ParsedItemsAnalysisData.cs
Comment (ladipro) [reply src/Build/BuildCheck/OM/ParsedItemsAnalysisData.cs]: It is not. Fixed!

File: src/Build/BuildCheck/Infrastructure/BuildEventsProcessor.cs
Comment (ladipro) [reply src/Build/BuildCheck/Infrastructure/BuildEventsProcessor.cs:79]: Done, thank you!

**Comment by ladipro** [comment]
Thank you for the quick review.

> Btw. - don't you want to extend the TaskParameterEventArgs as part of this PR? If not - would be nice to have addition PR for that - possibly merged prior this one.

The fact that assigning task outputs to properties is currently logged as text messages will need to be addressed in a careful and viewer-friendly way. Let me think about it a bit. I would prefer to do it in a separate PR without blocking this one. It should not be an issue for the double writes analyzer, for example.

File: src/Build/BuildCheck/Infrastructure/BuildCheckManagerProvider.cs
Comment (JanKrivanek) [reply src/Build/BuildCheck/Infrastructure/BuildCheckManagerProvider.cs:143]: That's correct. But regardless of acquisition state - the custom analyzer registration just takes a factory method - so any implementation of BuildAnalyzer can be injected this way as well.

File: src/Build/BuildCheck/OM/ParsedItemsAnalysisData.cs
Comment (JanKrivanek) [reply src/Build/BuildCheck/OM/ParsedItemsAnalysisData.cs:57]: Yeah - now I realize the initialization of the AnalysisData.ProjectFilePath here is likely not correct - it allways needs to be the build project - basically the one to which we apply the editorconfig rules (as their are applied based on that member).

We can have projectA.csproj importing the contoso.targets, where the later contains taskX task invocation. The TaskInvocationAnalysisData.ProjectFilePath must point to projectA.csproj (for the configuration correctness). But we as well want to have proper location info for reporting corectness - that should point to contoso.targets

**Comment by JanKrivanek** [comment]
> Thank you for the quick review.
> 
> > Btw. - don't you want to extend the TaskParameterEventArgs as part of this PR? If not - would be nice to have addition PR for that - possibly merged prior this one.
> 
> The fact that assigning task outputs to properties is currently logged as text messages will need to be addressed in a careful and viewer-friendly way. Let me think about it a bit. I would prefer to do it in a separate PR without blocking this one. It should not be an issue for the double writes analyzer, for example.

FYI @KirillOsenkov - just a heads up that another change is comming ;-) But I bet that move from textual to structured info is allways appreciated.

File: src/Build/BuildCheck/Infrastructure/BuildCheckManagerProvider.cs
Comment (ladipro) [reply src/Build/BuildCheck/Infrastructure/BuildCheckManagerProvider.cs:143]: At the risk of stating the obvious and not seeing a straightforward solution - the issue here is that `RegisterCustomAnalyzers` is an instance method on `BuildCheckManager` and there's no easy way to get to it from a test. I guess we could get to it through `BuildCheckManagerProvider.GlobalInstance` and a type cast. But it makes we wonder if we shouldn't expose programmatic custom analyzer registration publicly on e.g. `BuildParameters`, similar to loggers.

**Comment by KirillOsenkov** [comment]
Do we have a sample analyzer that would exercise this? It's always better to design APIs driven by a real-world use case. Would love to see how this API can be used at the consumer side.

**Comment by ladipro** [comment]
True that we don't currently have an analyzer that would need this. It is motivated by completeness - when something is interested in task parameters, it feels natural to provide both inputs and outputs.

The current model follows the MSBuild syntax: The viewer renders `Parameters` (input parameters), `OutputProperties` (output parameters that are assigned to props), and `OutputItems` (output parameters that are assigned to items). As an analyzer author I think I would expect the model to be task centric, i.e. I wouldn't so much be interested in where the output is going, but I would want to know the name of the output parameter so I can implement rules like "output parameter Foo of task Bar always has this particular shape". And again, the major issue is that the name "Foo" is currently not included anywhere. If a task assigns multiple output parameters to the same item, for example, the binlog is ambiguous.

**Review by YuliiaKovalova — APPROVED** [review]

File: src/Build/BuildCheck/Infrastructure/BuildCheckManagerProvider.cs
Comment (JanKrivanek) [reply src/Build/BuildCheck/Infrastructure/BuildCheckManagerProvider.cs:143]: I believe the type was recently test exposed by @YuliiaKovalova 

Overall the initial code wasn't quite component testing friendly and needs adjustments. Your current change or this suggestion are step forward in that direction, though just constructing in tests might be simplest solution.

Comment (JanKrivanek) [reply src/Build/BuildCheck/Infrastructure/BuildCheckManagerProvider.cs:143]: All is good - I misunderstood the scenario (e2e tests cannot use the instance of BuildCheckManager as it deosn't exist at the test setup time)

File: src/Build/BuildCheck/OM/ParsedItemsAnalysisData.cs
Comment (ladipro) [reply src/Build/BuildCheck/OM/ParsedItemsAnalysisData.cs]: I have pushed an update to fix this. The `ProjectFile` in event args really is the project being built. `TaskStartedEventArgs.TaskFile` is the file where the task is invoked from, which together with `LineNumber` and `ColumnNumber` makes a complete location information to be reported by build checks.

Comment (JanKrivanek) [reply src/Build/BuildCheck/OM/ParsedItemsAnalysisData.cs]: I'm blind :-) - but I do not see the location info (filename, line, col) - just the task assembly location

Comment (ladipro) [reply src/Build/BuildCheck/OM/ParsedItemsAnalysisData.cs]: https://github.com/dotnet/msbuild/blob/41f1bca43d750b2195f8357f5792a558500ddb25/src/Build/BuildCheck/OM/ParsedItemsAnalysisData.cs#L78

**Comment by ladipro** [comment]
Here's what I'm planning to do to address the missing output parameter name issue: https://github.com/ladipro/msbuild/pull/1/files
I will continue after this PR is merged.

File: src/BuildCheck.UnitTests/TaskInvocationAnalysisDataTests.cs
Comment (ladipro) [reply src/BuildCheck.UnitTests/TaskInvocationAnalysisDataTests.cs:144]: Follow-up work will be based on https://github.com/ladipro/msbuild/pull/1.

File: src/Build/BuildCheck/OM/ParsedItemsAnalysisData.cs
Comment (JanKrivanek) [reply src/Build/BuildCheck/OM/ParsedItemsAnalysisData.cs]: See - I told you I'm blind! :-D

Thanks ;-)

File: src/Build/BuildCheck/API/IBuildCheckRegistrationContext.cs
Comment (JanKrivanek) [reply src/Build/BuildCheck/API/IBuildCheckRegistrationContext.cs:12]: This feels nicely descriptive - I'd leave as is

File: src/Build/BuildCheck/OM/ParsedItemsAnalysisData.cs
Comment (JanKrivanek) [inline src/Build/BuildCheck/OM/ParsedItemsAnalysisData.cs:59]: Please add an xml comment here. Especially the `Value` might need explanation about possible types, scalar vs list values...

**Review by JanKrivanek — APPROVED** [review]
Thank you!!
:shipit:

**Review by YuliiaKovalova — APPROVED** [review]

**Pull request merged**
